Opentrons/opentrons
# refactor(app): Load custom labware from folder and display in app

## overview

This is a large PR for a large ticket. Closes #4244 by adding the following behind the `customLabware` feature flag:

- A custom labware directory in the app's preferences folder
- Periodic reading and parsing of JSON files in that folder
- A `labware` subtree in the app's state with custom labware files
- A simple UI for the `ListLabwareCard` based on that state

## changelog

- refactor(app): Load custom labware from folder and display in app

## review requests

**If possible, testing using a prod build from #builds is better than a dev build**

A lot of the bulk from this PR comes from tests! Almost all functionality added by this PR (components included) is covered, so give them a gander and don't get intimidated by the +1450

To test:

1. Open up your custom labware folder
    - Mac: `~/Library/Application Support/Opentrons/labware`
    - Windows: `%APPDATA%/Opentrons/labware`
    - Linux: `~/.config/Opentrons/labware`
2. Open up the app and, in the app's More settings, ensure:
    - Dev tools are enabled
    - `__DEV__ Custom Labware` is enabled
3. Go to "More > Custom Labware"

### checklist

- [ ] No custom labware shows up if your directory is empty
- [ ] If you add valid custom labware to the labware directory, they show up in the list
- [ ] If you add invalid custom labware, they do not show up
    - Unparseable .json file
    - JSON file that isn't labware
    - Opentrons standard labware
    - Two or more custom labware files that share the same [loadName, version, namespace]

### open design questions / out of scope

See the comments on Zeplin screen (linked in ticket above) for things that are not addressed by the design and thus out of scope for this review:

- Title of the card
- Missing column titles
- Missing filename
- Look of card with no custom labware in the folder
- Showing parseable but invalid labware in the list somehow

Also, the original ticket says the "app will watch custom labware directory". Given the size of this PR already and the lack (at this time) of designs / user flows that absolutely necessitate watching the directory, at the moment this PR **simply polls the folder while the list card is mounted**. We can make a watching ticket if/when that becomes a necessity
